projects/packages/backup-helper-script-manager/src/class-throw-on-errors.php

```diff
@@ -0,0 +1,497 @@
+
+namespace Automattic\Jetpack\Backup;
+
+use Exception;
+
+/**
+ * Wrappers for functions which throw an exception on errors and warnings instead of silently continuing operation.
+ *
+ * PHP is pretty lax with error reporting when doing I/O operations, e.g. a typical I/O helper function returns false,
+ * null, and/or emits a warning, but the whole PHP application continues operation. It's for the caller of the I/O
+ * helper function to test the returned result of the function, and notice + act upon I/O errors.
+ *
+ * We really want to know about each and every I/O error. Therefore, this class provides wrappers for some common
+ * (mostly) I/O operations that throw an exception on errors instead of just returning false, null, or something else.
+ * This wrapper class treats warnings as errors too.
+ *
+ * Given that static method names are similar to the ones used by PHP, they're prefixed with "t_" to not erroneously
+ * trigger various security scanners.
+ */
+class Throw_On_Errors {
+
+	/**
+	 * Execute a callable, throw an exception (together with a descriptive label) on PHP warnings / errors.
+	 *
+	 * @param callable $callable Callable to execute.
+	 * @param string   $label Label to add to the thrown exception to clarify what was attempted.
+	 *
+	 * @return mixed Callable's return value, if any.
+	 * @throws Exception On warnings thrown by the callable.
+	 * @noinspection PhpUnusedParameterInspection
+	 */
+	private static function throw_on_warnings( $callable, $label ) {
+		$old_error_reporting = error_reporting( - 1 );
+		$old_display_errors  = ini_set( 'display_errors', 'stderr' );
+
+		// phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_set_error_handler
+		set_error_handler(
+		/**
+		 * Temporary error handler.
+		 *
+		 * @see https://php-legacy-docs.zend.com/manual/php5/en/function.set-error-handler
+		 * @see https://www.php.net/manual/en/function.set-error-handler.php
+		 *
+		 * @param int $errno Level of the error raised.
+		 * @param string $errstr Error message.
+		 * @param string|null $errfile Filename that the error was raised in.
+		 * @param int|null $errline Line number where the error was raised.
+		 * @param array|null $errcontext Deprecated, unused.
+		 *
+		 * @return mixed
+		 * @throws Exception
+		 */
+			// phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+			function ( $errno, $errstr, $errfile = null, $errline = null, $errcontext = null ) {
+				throw new Exception( "$errstr (file: $errfile; line: $errline)" );
+			}
+		);
+
+		$result        = null;
+		$error_message = null;
+		if ( PHP_MAJOR_VERSION >= 7 ) {
+			// On PHP 7.x, all Exception and Error are subclasses of Throwable.
+			try {
+				$result = $callable();
+			} catch ( Throwable $throwable ) {
+				$error_message = $throwable->getMessage();
+			} catch ( Exception $exception ) {
+				$error_message = $exception->getMessage();
+			}
+		} else {
+			// On PHP 5.x, there's only Exception.
```

Comment:
We don't support 5.x anymore. The `composer.json` you're adding here specifically requires PHP >=7.0.

Even when we did, it never hurt to `catch ( Throwable )`, it would just never have caught anything.

Reply:
How come that [our plugin says that we do](https://wordpress.org/plugins/jetpack-backup/)?

<img width="349" alt="Screenshot 2023-12-12 at 14 02 10" src="https://github.com/Automattic/jetpack/assets/181949/461b083d-a471-47c5-b490-36b62c395987">

Reply:
This is because the Jetpack Backup plugin was last updated on September 20. Since then, multiple changes were made to the plugin, including a bump to the PHP version requirements:
https://github.com/Automattic/jetpack-backup-plugin/blob/trunk/CHANGELOG.md#23-alpha---unreleased

When you'll want to ship the changes you're introducing in this PR to customers, you'll need to ship a new update to the plugin. That will include your changes as well as all the ones currently listed as unreleased, so the plugin will require PHP 7.0+

Reply:
Thanks, removed PHP 5.x hacks in 202ce2b.

> When you'll want to ship the changes you're introducing in this PR to customers, you'll need to ship a new update to the plugin. That will include your changes as well as all the ones currently listed as unreleased, so the plugin will require PHP 7.0+

How do we do it? Is this something that Crew / Garage does, or should we ship that plugin ourselves somehow?

Reply:
You can release an update whenever you'd like, by following the steps described in PCYsg-KCk-p2
You can also ask for help in #jetpack-releases if you get stuck at any point.

Reply:
Thank you, we'll have a look!